### DISCUSSION TITLE
Context API

### DISCUSSION THREAD

**Discussion by lilnasy** [withastro/roadmap#731] [Proposal]
### Body

Exploration:
- [lilnasy/compiler#proposal-context](https://github.com/lilnasy/compiler/tree/proposal-context)
- [lilnasy/astro#proposal-context](https://github.com/lilnasy/astro/tree/proposal-context)

# Summary
A way for slotted Astro components to access data provided by a parent.

# Motivation
It is possible to pass data via props but that requires importing the receiving component directly. The context API is a solution that allows passing data to children components without having to import them directly.

# Example

### Overarching Component
```astro
---
import ButtonGroup from "../components/ButtonGroup.astro"
import Button      from "../components/Button.astro"
---
<ButtonGroup size="small">
    <Button/>
</ButtonGroup>
<ButtonGroup size="large">
    <Button/>
</ButtonGroup>
```

### Resulting HTML
_Formatted for clarity._
```html
<!DOCTYPE html>

<button-group>
    <button class="small"></button>
</button-group>

<button-group>
    <button class="large"></button>
</button-group>
```
### ButtonGroup.astro sets context
```astro
---
Astro.context.size = Astro.props.size
---
<button-group>
    <slot/>
</button-group>
```
### Button.astro reads context
```astro
---
const { size } = Astro.context
---
<button class={size}><slot/></button>
```

# Goals
- Improve composition story.
- Without breaking anything.

# Non goals
- Anything related to framework components. Scope is limited to astro components.
- Anything related to browser code. Scope is limited to server rendering.

**Comment by natemoo-re** [comment]
This looks great! I have two thoughts after looking at the initial implementation you've done here.

1. Would it be possible to avoid changing the component signature? I wonder if there's another way to use `$$result` (the first argument) to store contextual information. But on second thought, the way you've done it is much more explicit and doesn't require async tracking...
2. Language tools are a valid concern here. As proposed, this would likely rely on a global `Astro.context` interface. But in practice, only some components will have a specific context available, making every item in `Astro.context` optional. React's `useContext(Value)` API is actually quite an elegant solution to that problem.

**Reply by lilnasy** [reply]
1. The context argumment could be made optional, but I don't see a path without it. Also brings into question if we want to depend on AsyncLocalStorage, it is on most non-node server runtimes but not universal. Standard async context has a long road ahead as well. 
2. I agree, that is a concern. I'm curious if in practice, the context properties will be in the layout, or at least the defaults. `Astro.createContext()` could be a problematic pattern. How would users pass around the context key or type between components?

**Comment by mb21** [comment]
We're currently using middlewares and `Astro.locals` as a kind of context. Of course, this only works for global context for all Astro components (not a subtree like this context API here proposes) and for all routes (because middlewares run for all routes). Currently that's all we need, but I can see this being useful i the future.

What is a bit annoying to us already now though, is that we have to pass in `Astro.locals` as a prop in all our framework components and then initialize the client-side global state from that (either React/Solid Context or nanostore/shared-signal)...

**Comment by SantoJambit** [comment]
Just came here looking for some sort of context api. Aside from [another discussion](https://github.com/withastro/roadmap/discussions/445), I found [this library](https://withastro-utils.github.io/docs/guides/context/), but that modifies  Astro.locals, which is always the same object for all components within one request, so you could provide the value in one branch of the tree and read it in a completely different branch of the tree.

The API would have to allow having an object that gets extended without modifying the original and then pass the extended one to children.

I've done something similar with [tsx-dom-ssr](https://lusito.github.io/tsx-dom/tsx-dom-ssr/passing-props-deeply.html#createContext).

The basic idea in tsx-dom-ssr is like this:

```ts
export const SomeNumber = createContext({
  fallback: 10,
  description: "Some Number Context",
});
```

```tsx
<div>
  <SomeNumber.Provider value={7}>
    <Child />
  </SomeNumber.Provider>
  <SomeNumber.Provider value={42}>
    <OtherChild />
  </SomeNumber.Provider>
</div>
```

components could then use:
```astro
---
const someNumber = SomeNumber.for(Astro);
---
<div>{someNumber}</div>;
```

In contrast to the OP suggestion, this has the advantage, that you can have multiple providers within one astro component, without requiring a wrapper component to provide the value.

The code for it [isn't all that difficult](https://github.com/Lusito/tsx-dom/blob/master/packages/tsx-dom-ssr/src/createContext.ts).

The pseudo-code might work something like this:

```ts
export function createContext(options) {
    const type = Symbol(options.description);

    return {
        Provider: {
            // Not sure if there is something like an object with a render-function in astro. This is just an example.
            render(context, props, slots..) => {
                 const childContext = {
                         ...context,
                         providedValues: {
                              ...context.providedValues,
                              [type]: props.value
                         }
                 }
                 renderSlots(childContext, slots);
            }
        ),
        for(astro) {
            if (type in astro.providedValues) {
                return astro.providedValues[type];
            }

            if (options.fallback === undefined) {
                throw new Error(`Could not find Provider for "${String(type)}" and no fallback was configured!`);
            }

            return options.fallback;
        },
    };
}
```

This will probably not work with server islands though.

Since the names `context` and `createContext` are already used inside of the code, it would need new names that don't conflict here. Just spitballing here: `createProvidedValue`, `createLocalContext`, `createContextualValue`

I could imagine a different API that works like this:

```astro
---
const currentValue = Astro.getLocalContext("someNumber");
---
<div>
  <context type="someNumber" value={currentValue + 10}>
    <Child />
  </context>
  <context type="someNumber" value={42}>
    <OtherChild />
  </context>
</div>
```

And then defining the context types similar to content-collections with a schema definition. That would require more work on the IDE Plugins though I imagine.

**Comment by gVguy** [comment]
Bump. It would be great if Astro implemented this feature!

**Reply by delucis** [reply]
It’s not exactly a context API, but you can use `locals` for this in many cases: https://docs.astro.build/en/reference/api-reference/#locals

**Comment by 5cover** [comment]
This feature would come in handy to make relative headings : a Section component  increments the current level from the DOM context, and a Heading component uses this level to render h1...h6. This enables reusing headings in components without  having to count heading levels.

**Comment by tvald** [comment]
I've implemented a simple, STATIC RENDER ONLY library with the React-style `createContext()` / `useContext()` API and a couple of helpers to make it nice to use with Astro:

https://github.com/tvald/astro-static-context

```typescript
// theme-context.ts
import { createContext } from "astro-static-context";

export const ThemeContext = createContext({ theme: "light" });


// Parent.astro
---
import Child from "./Child.astro";
import { ThemeContext } from "./theme-context";
---

<ThemeContext value={{ theme: "dark" }}>
  <Child />
</ThemeContext>


// Child.astro
---
import { useContext } from "astro-static-context";
import { ThemeContext } from "./theme-context";

const { theme } = useContext(ThemeContext);
---

<p>Current theme: {theme}</p>
```

## Caveats
- ONLY works for static renders in node. This approach is unlikely to work with islands or in browsers.
- NOT reactive, does not cause a re-render of the subtree when values change.
- Exactly matches the React API naming for familiarity as a proof-of-concept, although this conflicts with existing Astro naming, as noted above in this thread. The API surface should be renamed appropriately if this were to be incorporated into the core framework.
